sql/src/influxql/provider.rs

```diff
@@ -0,0 +1,466 @@
+                    None
+                } else {
+                    let influx_type_res =
+                        map_column_type_to_influx_column_type(col, col_idx == timestamp_key_idx);
+                    Some(influx_type_res.map(|influx_type| (influx_type, arrow_col.clone())))
+                }
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Self {
+            columns: influx_columns,
+        })
+    }
+}
+
+impl InfluxSchema for InfluxSchemaImpl {
+    fn columns(&self) -> Vec<(InfluxColumnType, &ArrowField)> {
+        self.columns
+            .iter()
+            .map(|(influx_type, field)| (*influx_type, field))
+            .collect()
+    }
+
+    fn tags(&self) -> Vec<&ArrowField> {
+        self.columns
+            .iter()
+            .filter_map(|(influx_type, field)| {
+                if matches!(influx_type, InfluxColumnType::Tag) {
+                    Some(field)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    fn fields(&self) -> Vec<&ArrowField> {
+        self.columns
+            .iter()
+            .filter_map(|(influx_type, field)| {
+                if matches!(influx_type, InfluxColumnType::Field(..)) {
+                    Some(field)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    fn time(&self) -> &ArrowField {
+        // Time column must exist, has checked it when building.
+        let time_column = self
+            .columns
+            .iter()
+            .find(|(influx_type, _)| matches!(influx_type, InfluxColumnType::Timestamp))
+            .unwrap();
+
+        &time_column.1
+    }
+
+    fn column(&self, idx: usize) -> (InfluxColumnType, &ArrowField) {
+        let (influx_type, field) = &self.columns[idx];
+
+        (*influx_type, field)
+    }
+
+    fn find_index_of(&self, name: &str) -> Option<usize> {
+        self.columns
+            .iter()
+            .enumerate()
+            .find(|(_, (_, field))| field.name() == name)
+            .map(|(index, _)| index)
+    }
+}
+
+fn map_column_type_to_influx_column_type(
+    column: &ColumnSchema,
+    is_timestamp_key: bool,
+) -> Result<InfluxColumnType> {
+    // Time column
+    if is_timestamp_key {
+        if column.name == "time" && !column.is_nullable {
+            Ok(InfluxColumnType::Timestamp)
+        } else {
+            BuildSchema {
+                msg: format!("invalid time column, column:{column:?}"),
+            }
+            .fail()
+        }
+    } else if column.is_tag {
+        // Tag column
+        if matches!(column.data_type, DatumKind::String) && column.is_nullable {
+            Ok(InfluxColumnType::Tag)
+        } else {
+            BuildSchema {
+                msg: format!("invalid tag column, column:{column:?}"),
+            }
+            .fail()
+        }
+    } else {
```

Comment:
I guess we should define two functions called `map_column_type_to_influx_time_column` and `map_column_type_to_influx_tag_column`, which accepts the parameters and return `Option<InfluxColumnType>`. With the two function, it may make the structure of this function more simple.